packages/venia-ui/lib/components/CategoryList/categoryTile.js

```diff
@@ -1,9 +1,13 @@
 import React from 'react';
 import { arrayOf, string, shape } from 'prop-types';
 import { mergeClasses } from '../../classify';
-import { Link, resourceUrl } from '@magento/venia-drivers';
+import { Link } from '@magento/venia-drivers';
 import defaultClasses from './categoryTile.css';
 import { useCategoryTile } from '@magento/peregrine/lib/talons/CategoryList/useCategoryTile';
+import Image from '../Image';
+
+// This value mirrors classes.image.
```

Comment:
😅 

The `CategoryTile` component has no idea what style properties are associated with `.image`. If there's a reasonable belief that `.image` will want to use this value to derive a style, it should provide this value as a custom property. We don't manually synchronize values between a component and its default stylesheet.

```jsx
const IMAGE_SIZES = ["5rem"]

const Image = props => {
  // sizes should be an array
  const { sizes } = props
  const sizesAttribute = sizes.join(",")

  // final size must be a width, not a media condition
  const finalSize = sizes[sizes.length - 1]
  const style = { "--size": finalSize }

  return (
    <img sizes={sizesAttribute} style={style} />
  )
}
```

Reply:
> We can't declare sizes in CSS
`sizes` is an attribute on `img`. Therefore I had to "mirror" some CSS values in JS. If these values ever changed, they'd need to be updated in both places. While I added comments where appropriate to try and fend this off, it feels gross 🤮.

Thank you for addressing this, lol. I can change these to custom properties.

Reply:
I think this will work for single value `sizes` like this, but I'm struggling with the multiple sizes use case, specifically this line:

```js
// final size must be a width, not a media condition
const finalSize = sizes[sizes.length - 1]
```

We don't want to use that width if we have a media condition that is satisfied. 

Consider `IMAGE_SIZES = ["(max-width: 640px) 2rem", "5rem"];` and CSS that looks like:

```css
.image {
  width: 5rem;
}

@media (max-width: 640px) {
    .image {
        width: 2rem;
    }
}
```

If both of these values were re-written to use a `--size` custom property, one of them would be wrong:

```css
.image {
  width: var(--size); /* 5rem */
}

@media (max-width: 640px) {
    .image {
        width: var(--size); /* 5rem, but should be 2rem! */
    }
}
```

So now I'm thinking we need a custom property for each media condition. And as long as we're there, we also need one for the breakpoint(s) (`640px`) to keep that in sync.

So our CSS becomes:

```css
.image {
  width: var(--size-large); /* 5rem */
}

@media (max-width: var(--size-breakpoint)) {
    .image {
        width: var(--size-small); /* 2rem */
    }
}
```

Which already feels like way too much presentational logic in the JS, but that aside: who names these things?

It makes sense that a `component.js` would want to define the names the custom properties used by `component.css`, but how does it pass the `Image` component those names?

The `Image` component itself could also do it but then the custom property names couldn't have much meaning and would have to be things like `--size0`, `--size1`, `--size-breakpoint0` or similar.

Given `IMAGE_SIZES = ["(max-width: 640px) 2rem", "5rem"];`, `Image` could generate and set values for:
`--size0`: '5rem'
`--size-breakpoint1`: '640px'
`--size1`: '2rem'

But then `component.css` would look like:

```css
.image {
  width: var(--size0); /* 5rem */
}

@media (max-width: var(--size-breakpoint1)) {
    .image {
        width: var(--size1); /* 2rem */
    }
}
```

which is clunky.

I'd love some advice / feedback here.

Reply:
> Which already feels like way too much presentational logic in the JS, but that aside: who names these things?
> 
> It makes sense that a `component.js` would want to define the names the custom properties used by `component.css`, but how does it pass the `Image` component those names?
> 
> The `Image` component itself could also do it but then the custom property names couldn't have much meaning and would have to be things like `--size0`, `--size1`, `--size-breakpoint0` or similar.

Good questions.

**Who names things?**

`Image` would name them, not the component that rendered `Image`. Ordinarily, `Image` wouldn't know much about its elements' presentation, including how that presentation would change in response to media queries. The CSS would declare presentation, and `Image` would blindly apply that CSS.

In this case, size information is actually **data** , since it determines which image resource is requested. Since `Image` can presume that its presentation will depend on this data, in the same way an accordion's presentation depends on an `expanded` state, we need to relay this data to the stylesheet. The mechanics of how that happens are the responsibility of `Image`.

`Image` decides the custom property names in the same way it decides what classname keys it will apply to its elements. If my custom component, `Foo`, happens to render an `<Image />` element, it may supply overrides for the classname keys `Image` has defined, such as `classes.root`, but even if it were to supply new custom keys such as `classes.bar`, `Image` would not (*could not*) apply them to one of its internal elements.

Custom properties work the same way: if `Image` has logic to apply certain properties via `style`, with values derived from `props.sizes`, then `Foo` may adjust those property values via `props.sizes`. In other words, that's `Image`'s interface.

**Isn't this clunky?**

Less clunky than hoping things stay manually synced. 😅 

It's a bit verbose because you can't supply a collection of values via a single custom property—and even if you could, you couldn't have a rule *select* a single value from that collection. So we do have to supply multiple properties.

Reply:
Resolved in https://github.com/magento/pwa-studio/pull/1871/commits/ab27aa394627cf9292a4b133f4359a39f5a587fc.